File: src/main/java/dev/salmon/keystrokes/hud/api/KeyStyle.java
package dev.salmon.keystrokes.hud.api;

import cc.polyfrost.oneconfig.config.core.OneColor;
import cc.polyfrost.oneconfig.utils.color.ColorUtils;
import dev.salmon.keystrokes.config.KeystrokesConfig;

public class KeyStyle {
    public static final KeyStyle INSTANCE = new KeyStyle();

    public int interpolateColor(int toColor, int fromColor, float percent) {
        float avgRed = (toColor >> 16 & 0xFF) * percent + (fromColor >> 16 & 0xFF) * (1.0F - percent);
        float avgGreen = (toColor >> 8 & 0xFF) * percent + (fromColor >> 8 & 0xFF) * (1.0F - percent);
        float avgBlue = (toColor & 0xFF) * percent + (fromColor & 0xFF) * (1.0F - percent);
        float avgAlpha = (toColor >> 24 & 0xFF) * percent + (fromColor >> 24 & 0xFF) * (1.0F - percent);
        return ColorUtils.getColor(avgRed / 255.0F, avgGreen / 255.0F, avgBlue / 255.0F, avgAlpha / 255.0F);
    }

    public int backgroundColor(boolean pressed, float fade) {
        int thisColor = (pressed ? getBgPressed() : getBgUnpressed()).getRGB();
        if (fade < 1.0F) {
            int lastColor = (pressed ? getBgUnpressed() : getBgPressed()).getRGB();
            return interpolateColor(thisColor, lastColor, fade);
        }
        return thisColor;
    }

    public int textColor(boolean pressed, float fade) {
        int thisColor = (pressed ? getTextPressed() : getTextUnpressed()).getRGB();
        if (fade < 1.0F) {
            int lastColor = (pressed ? getTextUnpressed() : getTextPressed()).getRGB();
            return interpolateColor(thisColor, lastColor, fade);
        }
        return thisColor;
    }

    public OneColor getBgUnpressed() {
        return KeystrokesConfig.keystrokesElement.bgUnpressed;
    }

    public OneColor getBgPressed() {
        return KeystrokesConfig.keystrokesElement.bgPressed;
    }

    public OneColor getTextUnpressed() {
        return KeystrokesConfig.keystrokesElement.textUnpressed;
    }

    public OneColor getTextPressed() {
        return KeystrokesConfig.keystrokesElement.textPressed;
    }

    public boolean isShadowIdle() {
        return KeystrokesConfig.keystrokesElement.shadowIdle;
    }

    public boolean isShadowActive() {
        return KeystrokesConfig.keystrokesElement.shadowActive;
    }

    public boolean isRounded() {
        return KeystrokesConfig.keystrokesElement.rounded;
    }

    public int getCornerRadius() {
        return KeystrokesConfig.keystrokesElement.cornerRadius;
    }

    public int getFadingTime() {
        return KeystrokesConfig.keystrokesElement.fadingTime;
    }
}

File: src/main/kotlin/org/polyfrost/keystrokes/hud/api/KeyStyle.kt
package org.polyfrost.keystrokes.hud.api

import cc.polyfrost.oneconfig.config.core.OneColor
import cc.polyfrost.oneconfig.utils.color.ColorUtils
import org.polyfrost.keystrokes.config.KeystrokesConfig

object KeyStyle {
    fun interpolateColor(toColor: Int, fromColor: Int, percent: Float): Int {
        val avgRed = (toColor shr 16 and 0xFF) * percent + (fromColor shr 16 and 0xFF) * (1.0f - percent)
        val avgGreen = (toColor shr 8 and 0xFF) * percent + (fromColor shr 8 and 0xFF) * (1.0f - percent)
        val avgBlue = (toColor and 0xFF) * percent + (fromColor and 0xFF) * (1.0f - percent)
        val avgAlpha = (toColor shr 24 and 0xFF) * percent + (fromColor shr 24 and 0xFF) * (1.0f - percent)
        return ColorUtils.getColor(avgRed / 255.0f, avgGreen / 255.0f, avgBlue / 255.0f, avgAlpha / 255.0f)
    }

    fun backgroundColor(pressed: Boolean, fade: Float): Int {
        val thisColor = (if (pressed) this.bgPressed else this.bgUnpressed).getRGB()
        if (fade < 1.0f) {
            val lastColor = (if (pressed) this.bgUnpressed else this.bgPressed).getRGB()
            return interpolateColor(thisColor, lastColor, fade)
        }
        return thisColor
    }

    fun textColor(pressed: Boolean, fade: Float): Int {
        val thisColor = (if (pressed) this.textPressed else this.textUnpressed).getRGB()
        if (fade < 1.0f) {
            val lastColor = (if (pressed) this.textUnpressed else this.textPressed).getRGB()
            return interpolateColor(thisColor, lastColor, fade)
        }
        return thisColor
    }

    val bgUnpressed: OneColor
        get() = KeystrokesConfig.keystrokesElement.bgUnpressed

    val bgPressed: OneColor
        get() = KeystrokesConfig.keystrokesElement.bgPressed

    val textUnpressed: OneColor
        get() = KeystrokesConfig.keystrokesElement.textUnpressed

    val textPressed: OneColor
        get() = KeystrokesConfig.keystrokesElement.textPressed

    val isShadowIdle: Boolean
        get() = KeystrokesConfig.keystrokesElement.shadowIdle

    val isShadowActive: Boolean
        get() = KeystrokesConfig.keystrokesElement.shadowActive

    val isRounded: Boolean
        get() = KeystrokesConfig.keystrokesElement.rounded

    val cornerRadius: Int
        get() = KeystrokesConfig.keystrokesElement.cornerRadius

    val fadingTime: Int
        get() = KeystrokesConfig.keystrokesElement.fadingTime
}